ath79: mikrotik: add support for early RB911 Lite5 NAND boards

rogerpueyo opened the pull request:
These two patches add support for the early version of the MikroTik RB911 Lite5 boards with NAND storage. They were very almost hardware-identical to the already supported SXT Lite5, even the bootloader passes the same board identifier "sxt5n". The only difference is found in the tiny SPI NOR storage for the bootloader, which is half the size (64 kB for RB911Lite5NAND, 128 kB for the SXTLite5).

The first patch adapts the "SXT 5N" .dtsi to handle both 64 kB and 128 kB (bootloader2 partition will be added on a per-device basis, when needed). The second patch adds actual support for the RB911Lite5NAND boards.

Later versions of the RB911Lite5, which used an SPI NOR chip as their main storage, are not supported by the "sxt5n" platform, but by the "911L" one. The whole discussion and findings can be found at https://bugs.openwrt.org/index.php?do=details&task_id=3471.

rogerpueyo committed:
ath79: mikrotik: prepare SXT 5N platform for 64/128 kB SPI

The MikroTik SXT 5N (sxt5n) platform is commonbase for different devices with
very similar hardware. Some of them have a 128 kB SPI NOR flash (e.g., SXT
5nD r2) while others have a smaller 64 kB chip (early 911 Lite5 models with
NAND storage).

This patch makes the dtsi file compatible with both memory sizes, and lets
adding the bootloader2 partition on a per-device basis.

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

rogerpueyo committed:
ath79: mikrotik: add support for early RB9111 Lite5 NAND boards

The MikroTik Lite5 (product code RB911-5Hn) is a small CPE-type board with an
integrated 5 GHz single chain radio using the AR9344 SoC.

The early versions of this model were based on the "sxt5n" platform and had a
NAND flash storage. Later versions of the device kept the same product code,
but switched to an SPI NOR memory and were based on the "911L" platform. This
patch adds support only for the early version.

Specifications:
 - SoC: Atheros AR9344
 - RAM: 64 MB
 - Storage: 128 MB NAND
 - Wireless: Atheros AR9340 (SoC) 802.11a/n 1x1:1
 - Ethernet: Atheros AR8229 switch (SoC), 1x 10/100 port, 8-32 Vdc PoE in
 - 6 user-controllable LEDs:
  · 1x power (blue)
  · 1x wlan (green)
  · 4x rssi (green)

 See https://mikrotik.com/product/RB911-5Hn for more details (notice the text
 states the storage size is 16 MB, but the picture clearly shows a device with
 a 128 MB NAND flash chip).

Notes:
 The device was supported in the ar71xx target, as it had almost identical
 hardware as the SXT Lite 5 (see 3519322e6238913eeae717a598c9760ddf40436c).
 The only difference is the missing buzzer, which has been moved from
 the common .dtsi file to the device-specific SXT Lite 5 dts.

Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>
Tested by: Alexandros C. Couloumbis
Signed-off-by: Roger Pueyo Centelles <roger.pueyo@guifi.net>

adschm commented on target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:
```diff
@@ -6,3 +6,11 @@
 };
+
+&routerboot {
+	bootloader2: partition@10000 {
```
Is the bootloader2: DT label needed for something I don't see, or can it be dropped?

rogerpueyo replied:
Dropped, thanks.

f00b4r0 replied:
Don't do this. Don't split partitions between dtsi and dts. This is going to be a maintenance nightmare.
Why do you need to do this to begin with? The 64kB devices do have a secondary bootloader.

adschm commented:
Has this early device been sold regularly?

Note that the Signed-by in your second patch is duplicated.

adschm commented on target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds:
```diff
@@ -10,6 +10,7 @@ case "$board" in
 mikrotik,routerboard-lhg-2nd)
 	ucidef_set_led_netdev "lan" "lan" "green:lan" "eth0"
 	;;
+mikrotik,routerboard-911-5hn-nand|\
```
Block needs to be moved to keep sorting.

rogerpueyo replied:
True. Thanks.

adschm commented:
So, am I understanding this right:
The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand" and "spi" as you seem to do it? So, early=nand and late=spi-only ...?

adschm commented on target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5n.dtsi:
```diff
@@ -99,9 +99,9 @@
+			routerboot: partition@0 {
 				label = "RouterBoot";
-				reg = <0x0 0x20000>;
+				reg = <0x0 0x0>;
```
@f00b4r0 Is setting the size to 0x0 here harmless? The last partition for a 64k chip would be soft_config then which has no size as well ...

f00b4r0 replied:
now, this is invalid. The top-level "Routerboot" partition is a "fixed-partitions" style partition and its start offset and size must be correctly set.

tmn505 replied:
As @f00b4r0 said, fixed means no size auto-detection guaranteed ever, so it could break in the future. For reference the flash map:
````
dev:    size   erasesize  name
mtd0: 00040000 00020000 "booter"
mtd1: 003c0000 00020000 "kernel"
mtd2: 07c00000 00020000 "ubi"
mtd3: 00010000 00001000 "RouterBoot"
mtd4: 0000c000 00001000 "bootloader1"
mtd5: 00001000 00001000 "hard_config"
mtd6: 00001000 00001000 "bios"
mtd7: 00001000 00001000 "soft_config"
````

rogerpueyo replied:
Actual size for each device is set now.

adschm commented:
You amended the change to the wrong commit.

rogerpueyo commented:
> Note that the Signed-by in your second patch is duplicated.

Removed. Thanks.

> Has this early device been sold regularly?

I don't actually own the device. From the discussion we had at https://bugs.openwrt.org/index.php?do=details&task_id=3471 I understand it was sold regularly and, at some point, MikroTik slightly changed the hardware (just the storage, it seems) but kept selling it under the same name. This website from 2014 states it has NAND storage:

f00b4r0 reviewed (changes requested):
Replying to @adschm's ping.

Please note once again that supporting NAND devices is _BROKEN_ in current OpenWRT and will eventually lead to bricked devices.

f00b4r0 commented on target/linux/ath79/dts/ar9344_mikrotik_routerboard-sxt-5nd-r2.dts:
```diff
@@ -6,3 +6,11 @@
 	model = "MikroTik RouterBOARD SXT 5nD r2 (SXT Lite5)";
 };
+
+&routerboot {
+	partition@10000 {
+		label = "bootloader2";
+		reg = <0x10000 0x10000>;
+		read-only;
+	};
+};
```
Please don't do this. This makes no sense, and will be a maintenance nightmare. If the partition tables differ between devices, each DTS should have the full table I think. @adschm opinion?

adschm replied:
I don't have a problem with splitting in general. One can improve the situation by adding comments to the possibly "incomplete" setup in the DTSI. I'd be more cautious with the routerboot situation, though, since this harder to grasp than the simple list of partitions you typically have.

So, no, I don't reject this generally. However, with the answer on the other question with respect to the size, this option is dead anyway as the size requirement will also require to have two full definitions.

tmn505 replied:
Inevitably the NOR variant will be added, which will further clobber the visibility if kept like this. For readability full flash table would be preferable.

rogerpueyo replied:
OK, this was fixed, as each device got its partition table.

f00b4r0 commented:
> So, am I understanding this right:
> The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand" and "spi" as you seem to do it? So, early=nand and late=spi-only ...?

We're back onto the Mikrotik naming nightmare, #3270 all over again. I'm afraid this will eventually demonstrate that 1 image per board doesn't work for this vendor 😛

rogerpueyo commented:
> Note that the Signed-by in your second patch is duplicated.

Removed. Thanks.

> You amended the change to the wrong commit.

True. Fixed, thanks.

> Has this early device been sold regularly?

This MikroTik brochure (~2014) states 128 MB NAND: https://www.cyberbajt.pl/download/routerboard/911/RB911_series-140422133903.pdf
Also this website from 2014 states it has NAND storage: https://www.roc-noc.com/mikrotik/routerboard/RB911-5Hn.html

Looking at the wayback machine, at some point between Nov 2015 https://web.archive.org/web/20151118192818/http://routerboard.com/RB911-5Hn and Dec 2015 they had already switched to SPI NOR.

I don't actually own the device. From the discussion we had at https://bugs.openwrt.org/index.php?do=details&task_id=3471 I understand it was sold regularly and, at some point, MikroTik slightly changed the hardware (just the storage, it seems) but kept selling it under the same name.

> no, this is invalid. The top-level "Routerboot" partition is a "fixed-partitions" style partition and its start offset and size must be correctly set.
>
> Please don't do this. This makes no sense, and will be a maintenance nightmare. If the partition tables differ between devices, each DTS should have the full table I think. @adschm opinion?

So, one partition table per device. :+1:

> So, am I understanding this right:
> The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand"  and "spi" as you seem to do it? So, early=nand and late=spi-only ...?

Yes, "early" version has 64 kB SPI NOR + 128 MB NAND, "late" version has 16 MB SPI NOR. Both with the same model name. Glorious.

f00b4r0 commented:
> > The early version uses NAND and the later one doesn't. I.e., one could name the devices with "early" and "late", or with "nand"  and "spi" as you seem to do it? So, early=nand and late=spi-only ...?
> 
> Yes, "early" version has 64 kB SPI NOR + 128 MB NAND, "late" version has 16 MB SPI NOR. Both with the same model name. Glorious.

The problem with that naming scheme is that the average user will never know which applies to their device. So we'll probably need to maintain some table. Alternatively, I'd suggest e.g. "-nand128M" and "-nor16M" suffixes, they're quite more closely related to both the hardware and the way mikrotik presents their devices on their website. My 2c.

acoul commented:
Greetings,

using the above patches on a recent ath79/trunk on my recently purchased rb911-nand device, provided a successful working image

Linux version 5.4.81 (alex@aifnis) (gcc version 9.3.0 (OpenWrt GCC 9.3.0 r15132-9c13513ce1)) #0 Sun Dec 6 08:23:21 2020

CPU0 revision is: 0001974c (MIPS 74Kc)
MIPS: machine is MikroTik RouterBOARD 911-5Hn NAND (911 Lite5)
SoC: Atheros AR9344 rev 2

jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
nand: device found, Manufacturer ID: 0x98, Chip ID: 0xf1
nand: Toshiba NAND 128MiB 3,3V 8-bit
nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
3 fixed-partitions partitions found on MTD device ar934x-nand
Creating 3 MTD partitions on "ar934x-nand":
spi-nor spi0.0: mx25l512e (64 Kbytes)
1 fixed-partitions partitions found on MTD device spi0.0
Creating 1 MTD partitions on "spi0.0":

tmn505 commented on target/linux/ath79/image/mikrotik.mk:
```diff
@@ -9,6 +9,15 @@ define Device/mikrotik_routerboard-493g
+  $(Device/mikrotik_nand)
+  SOC := ar9344
+  DEVICE_MODEL := RouterBOARD 911-5Hn NAND (911 Lite5)
+  DEVICE_PACKAGES += rssileds kmod-gpio-beeper
```
There is no beeper on 911 board on both variants. Please remove package and move the dtsi node to appropriate dts.

rogerpueyo replied:
@adschm, @tmn505,

Thanks for the review.

> Alternatively, I'd suggest e.g. "-nand128M" and "-nor16M" suffixes,

> I wonder if image name shouldn't be changed to routerboard-911-xhn-nand, since 2Hn version will work with the same image.

I added the -nand128m suffix, and changed the name to mention the flash size. I also added the 2.4 GHz version 2Hn in `mikrotik.mk`:
```
define Device/mikrotik_routerboard-911-5hn-nand128m
  $(Device/mikrotik_nand)
  SOC := ar9344
  DEVICE_MODEL := RouterBOARD 911-5Hn NAND 128MB (911 Lite5)
  DEVICE_ALT0_VENDOR := MikroTik
  DEVICE_ALT0_MODEL := RouterBOARD 911-2Hn NAND 128MB (911 Lite2)
  DEVICE_PACKAGES += rssileds
  SUPPORTED_DEVICES += rb-911-5hn-nand
endef
TARGET_DEVICES += mikrotik_routerboard-911-5hn-nand128m
```

> There is no beeper on 911 board on both variants. Please remove package and move the dtsi node to appropriate dts.

I removed the beeper and moved the node to the SXT Lite 5 dts.

> Inevitably the NOR variant will be added, which will further clobber the visibility if kept like this. For readability full flash table would be preferable.
> As @f00b4r0 said, fixed means no size auto-detection guaranteed ever, so it could break in the future. For reference the flash map:

I created a whole SPI NOR partition table on each device.

> Additionally there is ethernet LED which works on ar71xx image but not with ath79.

Unfortunately, I can't verify this, since I don't own the device.

rogerpueyo replied:
Except for the Ethernet LED, which I can't test, I'd say everything else here is solved.

tmn505 reviewed (changes requested):
Additionally there is ethernet LED which works on ar71xx image but not with ath79.

I wonder if image name shouldn't be changed to routerboard-911-xhn-nand, since 2Hn version will work with the same image.

Thanks.

ynezz commented:
Ping.

acoul commented:
if anyone is willing to provide a suggested patch, I will be happy to test it & report back

f00b4r0 commented:
Besides the naming issue I have no further comment. I am however of the opinion that we shouldn't add more Mikrotik NAND boards given the broken-by-design support implementation.

acoul commented:
this board works wonderfully under (patched) ath79/trunk. I agree that MikroTik sucks in shamefully violating GPL again & again, but their hw combined with open source is quite worthy 

> root@OpenWrt:~# cat /proc/cpuinfo 
system type		: Atheros AR9344 rev 2
machine			: MikroTik RouterBOARD 911-5Hn NAND (911 Lite5)
processor		: 0
cpu model		: MIPS 74Kc V4.12
BogoMIPS		: 299.26
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 32
extra interrupt vector	: yes
hardware watchpoint	: yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa			: mips1 mips2 mips32r1 mips32r2
ASEs implemented	: mips16 dsp dsp2
Options implemented	: tlb 4kex 4k_cache prefetch mcheck ejtag llsc dc_aliases perf_cntr_intr_bit perf
shadow register sets	: 1
kscratch registers	: 0
package			: 0
core			: 0
VCED exceptions		: not available
VCEI exceptions		: not available

> root@OpenWrt:~# cat /proc/interrupts 
           CPU0       
  4:       1305      MIPS   4  19000000.eth
  7:     793290      MIPS   7  timer
  9:         20      MISC   3  ttyS0
 12:     130675      INTC   0  ath9k
 21:      20886      MISC  21  ar934x-nand
ERR:         24

> root@OpenWrt:~# df
Filesystem           1K-blocks      Used Available Use% Mounted on
/dev/root                 2304      2304         0 100% /rom
tmpfs                    29856        60     29796   0% /tmp
/dev/ubi0_2             108788       932    103020   1% /overlay
overlayfs:/overlay      108788       932    103020   1% /
tmpfs                      512         0       512   0% /dev

adschm commented on target/linux/ath79/mikrotik/base-files/etc/board.d/01_leds:
```diff
@@ -23,6 +24,9 @@ mikrotik,routerboard-sxt-5nd-r2)
 	ucidef_set_led_rssi "rssimediumhigh" "rssimediumhigh" "green:rssimediumhigh" "wlan0" "60" "100"
 	ucidef_set_led_rssi "rssihigh" "rssihigh" "green:rssihigh" "wlan0" "80" "100"
 	;;
+mikrotik,routerboard-lhg-2nd)
```
Unrelated change?

rogerpueyo replied:
Weird. Maybe I screwed it during a rebase, the board is already configured in line ~9. Removing it.

adschm commented:
If I understand correctly, the 0x0 size issue is not resolved yet either.

rogerpueyo commented:
> If I understand correctly, the 0x0 size issue is not resolved yet either.

Just fixed it. 0x0 is now 0x20000 for the SXT5 (128 kB), 0x10000 for the 911Lite5.

Rebased.

tmn505 reviewed (approved):
Tested the PR on both kernels and it's working fine. The eth LED is not mandatory, I also don't know yet how to enable it. The device name could be tuned as this image will also work for 2Hn and aside from 128 MiB NAND variant there's no other, so nand128 could be shortened to nand. I'll leave the decision of a name for the device to someone who'll be merging this.

rogerpueyo commented:
> The device name could be tuned as this image will also work for 2Hn and aside from 128 MiB NAND variant there's no other, so nand128 could be shortened to nand. I'll leave the decision of a name for the device to someone who'll be merging this.

I don't have a strong opinion on this neither, both options are fine for me. :+1: 

Rebased current master which conflicted with the patch.
